src/wasm2asm.h

```diff
@@ -196,7 +197,116 @@ class Wasm2AsmBuilder {
+                                             b.makeConst(Literal(1))));
+  Binary* subExp =
+      b.makeBinary(subOp,
+                   b.makeConst(is32Bit ? Literal(32 - 1) : Literal(64 - 1)),
+                   b.makeUnary(clzOp, xorExp));
+  If* body = b.makeIf(b.makeUnary(eqzOp, b.makeGetLocal(0, i32)),
+                      b.makeConst(is32Bit ? Literal(32) : Literal(64)),
+                      subExp);
+  return b.makeFunction(funcName,
+                        std::vector<NameType>{NameType("x", argType)},
+                        argType,
+                        std::vector<NameType>{},
+                        body);
+}
+
+Function* makePopcntFunc(MixedArena& allocator, bool is32Bit=true) {
+  Builder b(allocator);
+  // int c; for (c = 0; x != 0; c++) { x = x & (x - 1) }; return c
+  Name funcName = is32Bit ? Name(POPCNT32) : Name(POPCNT64);
+  BinaryOp addOp = is32Bit ? AddInt32 : AddInt64;
+  BinaryOp subOp = is32Bit ? SubInt32 : SubInt64;
+  BinaryOp andOp = is32Bit ? AndInt32 : AndInt64;
+  UnaryOp eqzOp = is32Bit ? EqZInt32 : EqZInt64;
+  WasmType argType = is32Bit ? i32 : i64;
+  Name loopName("l");
+  Name blockName("b");
+  Break* brIf =
+      b.makeBreak(blockName,
+                  b.makeGetLocal(1, i32),
+                  b.makeUnary(eqzOp, b.makeGetLocal(0, argType)));
+  SetLocal* update =
+      b.makeSetLocal(0, b.makeBinary(andOp,
+                                     b.makeGetLocal(0, argType),
+                                     b.makeBinary(subOp,
+                                                  b.makeGetLocal(0, argType),
+                                                  b.makeConst(Literal(1)))));
+  SetLocal* inc =
+      b.makeSetLocal(1, b.makeBinary(addOp,
+                                     b.makeGetLocal(1, argType),
+                                     b.makeConst(Literal(1))));
+  Break* cont = b.makeBreak(loopName);
+  Loop* loop = b.makeLoop(loopName, b.blockify(brIf, update, inc, cont));
+  Block* loopBlock = b.blockifyWithName(loop, blockName);
+  SetLocal* initCount = b.makeSetLocal(1, b.makeConst(Literal(0)));
+  return b.makeFunction(funcName,
+                        std::vector<NameType>{NameType("x", argType)},
+                        argType,
+                        std::vector<NameType>{NameType("count", argType)},
+                        b.blockify(initCount, loopBlock));
+}
+Function* makeRotFunc(MixedArena& allocator, bool isLRot, bool is32Bit=true) {
+  Builder b(allocator);
+  // left rotate is:
+  // (((((~0) >>> k) & x) << k) | ((((~0) << (w - k)) & x) >>> (w - k)))
+  // where k is shift modulo w. reverse shifts for right rotate
+  static Name names[2][2] = {{Name(ROTR64), Name(ROTR32)},
+                             {Name(ROTL64), Name(ROTL32)}};
+  static BinaryOp shifters[2][2] = {{ShrUInt64, ShrUInt32},
+                                    {ShlInt64, ShlInt32}};
+  Name funcName = names[isLRot][is32Bit];
+  BinaryOp lshift = shifters[isLRot][is32Bit];
```

Comment:
Kind of interesting that `lshift` can be a right shift when `!isLRot`. That's probably fine, but a bit odd.